### ISSUE TITLE
Allow inspection of not-running containers

### ISSUE THREAD

**Issue by flavio** [SUSE/zypper-docker#136]
Currently zypper-docker can scan only running containers. It would be nice to allow it to analyze also the non-running ones.

bsc#1064996

**Comment by lightlace** [comment]
I would like to tackle this.

**Comment by flavio** [comment]
@lightlace: the bug entry is a bit messy as you can see. I would just focus on this workflow:

* User enters the ID of a stopped container
* zypper-docker figures out from which image it was started
* zypper-docker follows the usual workflow

Let's not do the `docker commit` and then inspect thing.

**Comment by cyphar** [comment]
Using the source image could lead to confusion -- at that point you're not actually scanning the files changed in the container.

**Comment by lightlace** [comment]
Well, think of the following scenario:

The user starts a container from an openSUSE 42.3 image and updates it manually.
Then this container is stopped.
When zypper-docker now gets the container ID, it would say that there are updates, but the stopped container is in fact already updated..

Or am i wrong with this assumption?

**Comment by cyphar** [comment]
Yes. Or the user has installed an additional package which wasn't in the image, which `zypper-docker` won't see at all.

**Comment by lightlace** [comment]
After discussing this issue, we came up with the following solution:

# Analyzing a running container

By default all the container related commands will:
  * print a warning message: this tells the user zypper-docker will inspect the image from which the container was started. Changes made to the running container are not going to be analyzed. Use the "--force" argument to handle also eventual packages installed inside of the running container.
  * inspect the image from which the container was started. That's the current behavior
  
# Analyzing a running container with the "--force" flag

* commit the running container to a temporary image
* inspect that image
* remove the temporary image
* provide some feedback about what is being done

# Analyze a stopped container

* commit the running container to a temporary image
* inspect that image
* remove the temporary image
* provide some feedback about what is being done

**Comment by lightlace** [comment]
In order to commit a stopped container to a temporary image, i came up with the following sequence:

- the Container ID is given to zypper-docker via user input
- look for the image name by using docker inspect
- parse the image name in order to get the repo and the tag, which is necessary for committing the container to a new image
- commit the container to an image

I ran into the following problem:
The function parseImageName (helpers.go, line 176) needs the image name in the following order:
**repo/image:tag**
But the repo name can't be determined by docker inspect, so i can't commit the container to the image.

Any ideas? @flavio @cyphar @vrothberg @mssola

**Comment by cyphar** [comment]
The new image doesn't need to have a name. `docker commit $some_container` (without any image argument) will create a new image that can be referenced by the image ID provided. You don't need to do parsing of `docker inspect` at all. It would just look like this:

```
% docker commit $some_container
sha256:blah...
% docker run --rm blah... zypper --whatever-we-do-here
... output ...
% docker rmi blah...
```

**Comment by vrothberg** [comment]
I agree with @cyphar. Another benefit from not providing a name is that the image tag/ID of the created image will be unique (Docker takes care of that). In terms of code, this means to leave tag/ID/name be an empty string.

**Comment by lightlace** [comment]
the commands are going to be executed via exec.Command, am i right?

**Comment by vrothberg** [comment]
> the commands are going to be executed via exec.Command, am i right?

I prefer using the golang engine API. We're still using the deprecated one in zypper-docker, but this allows us proper error handling and it will save some resources.

**Comment by flavio** [comment]
I agree with what the others said. Don't provide neither a repository nor a tag, leave them empty and just use the image ID automatically created by docker.

Also stick with using the Docker API, don't rely on exec.

**Comment by cyphar** [comment]
My `docker ...` examples were just an example of what the logic would be, not trying to say that we should use `exec.Command`. :wink:
